python/dgl/transforms/module.py

```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
+            1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
+            1.0000, 1.0000])
+
+    Case2: Normalize features of a heterogeneous graph.
+
+    >>> transform = NormalizeFeatures()
+    >>> g = dgl.heterograph({
+    ...     ('user', 'follows', 'user'): (torch.tensor([1, 2]), torch.tensor([3, 4])),
+    ...     ('player', 'plays', 'game'): (torch.tensor([2, 2]), torch.tensor([1, 1]))
+    ... })
+    >>> g.ndata['h'] = {'game': torch.randn(2, 5), 'player': torch.randn(3, 5)}
+    >>> print(g.ndata['h']['game'].sum(1), g.ndata['h']['player'].sum(1))
+    tensor([ 4.4201, -4.0683]) tensor([-2.2460, -1.1204, -2.0254])
+    >>> g.edata['w'] = {('user', 'follows', 'user'): torch.randn(2, 5), ('player', 'plays', 'game'): torch.randn(2, 5)}
+    >>> print(g.edata['w'][('user', 'follows', 'user')].sum(1), g.edata['w'][('player', 'plays', 'game')].sum(1))
+    tensor([-1.2663,  3.3789]) tensor([4.1371, 1.4743])
+    >>> new_g = transform(g)
+    >>> print(new_g.ndata['h']['game'].sum(1), new_g.ndata['h']['player'].sum(1))
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000, 1.0000])
+    >>> print(new_g.edata['w'][('user', 'follows', 'user')].sum(1), new_g.edata['w'][('player', 'plays', 'game')].sum(1))
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000])
+    """
+    def __init__(self, node_feat_names=None, edge_feat_names=None):
+        self.node_feat_names = node_feat_names
+        self.edge_feat_names = edge_feat_names
+
+    def __call__(self, g):
+        def normalize(feat):
+            feat = feat - feat.min()
+            feat.div_(feat.sum(dim=-1, keepdim=True).clamp_(min=1.))
+            return feat
+
+        if self.node_feat_names is None:
+            self.node_feat_names = g.ndata.keys()
+
+        if self.edge_feat_names is None:
+            self.edge_feat_names = g.edata.keys()
+
+        for node_feat_name in self.node_feat_names:
+            if isinstance(g.ndata[node_feat_name], torch.Tensor):
+                g.ndata[node_feat_name] = normalize(g.ndata[node_feat_name])
+            else:
+                for ntype in g.ndata[node_feat_name].keys():
+                    g.nodes[ntype].data[node_feat_name] = normalize(g.nodes[ntype].data[node_feat_name])
+
+        for edge_feat_name in self.edge_feat_names:
+            if isinstance(g.edata[edge_feat_name], torch.Tensor):
+                g.edata[edge_feat_name] = normalize(g.edata[edge_feat_name])
+            else:
+                for etype in g.edata[edge_feat_name].keys():
+                    g.edges[etype].data[edge_feat_name] = normalize(g.edges[etype].data[edge_feat_name])
+
+        return g
+
+class NodeFeatureMasking(BaseTransform):
+    r"""Randomly mask node features, as described in `An Empirical Study of Graph Contrastive Learning
+    <https://arxiv.org/abs/2109.01116>`__.
+
+    Parameters
+    ----------
+    p : float, optional
+        Probability of masking a node feature. Default: `0.5`.
```

Comment:
Add an argument to specify the name of the feature(s) to mask?

Reply:
Fixed